File: elixir/list-ops/list_ops.exs
defmodule ListOps do
  # Please don't use any external modules (especially List) in your
  # implementation. The point of this exercise is to create these basic functions
  # yourself.
  #
  # Note that `++` is a function from an external module (Kernel, which is
  # automatically imported) and so shouldn't be used either.

  @spec count(list) :: non_neg_integer
  def count(list), do: do_count(list, 0)
  defp do_count([], total), do: total
  defp do_count([_ | tail], total) do
    do_count(tail, total + 1)
  end

  @spec reverse(list) :: list
  def reverse(list), do: do_reverse(list, [])
  defp do_reverse([], reversed), do: reversed
  defp do_reverse([head|tail], reversed) do
    do_reverse(tail, [head | reversed])
  end

  @spec map(list, (any -> any)) :: list
  def map(list, func), do: do_map(reverse(list), func, [])
  defp do_map([], _func, result), do: result
  defp do_map([head|tail], func, result) do
    do_map(tail, func, [func.(head) | result])
  end

  @spec filter(list, (any -> as_boolean(term))) :: list
  def filter(list, func), do: do_filter(reverse(list), func, [])
  defp do_filter([], _func, result), do: result
  defp do_filter([head|tail], func, result) do
    do_filter(
      tail,
      func,
      (if func.(head), do: [head | result], else: result)
    )
  end

  @type acc :: any
  @spec reduce(list, acc, ((any, acc) -> acc)) :: acc
  def reduce([], acc, _func), do: acc
  def reduce([head|tail], acc, func) do
    reduce(tail, func.(head, acc), func)
  end

  @spec append(list, list) :: list
  def append(a, b), do: do_append(reverse(a), b)
  defp do_append([], b), do: b
  defp do_append([head|tail], b) do
    do_append(tail, [head | b])
  end

  @spec concat([[any]]) :: [any]
  def concat(ll), do: do_concat(reverse(ll), [])
  defp do_concat([], result), do: result
  defp do_concat([head|tail], result) do
    do_concat(tail, append(head, result))
  end
end

File: go/list-ops/list_ops.go
package listops

// IntList is an abstraction of a list of integers which we can define methods on
type IntList []int

func (s IntList) Foldl(fn func(int, int) int, initial int) int {
	acc := initial
	for _, n := range s {
		acc = fn(acc, n)
	}
	return acc
}

func (s IntList) Foldr(fn func(int, int) int, initial int) int {
	acc := initial
	for i := len(s) - 1; i >= 0; i-- {
		acc = fn(s[i], acc)
	}
	return acc
}

func (s IntList) Filter(fn func(int) bool) IntList {
	filtered := IntList{}
	for _, n := range s {
		if fn(n) {
			filtered = append(filtered, n)
		}
	}
	return filtered
}

func (s IntList) Length() int {
	return len(s)
}

func (s IntList) Map(fn func(int) int) IntList {
	mapped := make(IntList, len(s))
	for i, n := range s {
		mapped[i] = fn(n)
	}
	return mapped
}

func (s IntList) Reverse() IntList {
	length := len(s)
	reversed := make(IntList, length)
	for i := range s {
		reversed[i] = s[length-i-1]
	}
	return reversed
}

func (s IntList) Append(lst IntList) IntList {
	out := IntList{}
	out = append(out, s...)
	out = append(out, lst...)
	return out
}

func (s IntList) Concat(lists []IntList) IntList {
	out := IntList{}
	out = append(out, s...)
	for _, list := range lists {
		out = append(out, list...)
	}
	return out
}
